microsoft/PowerToys
# Autoupdate: implement updating bootstrapper utility

## Summary of the Pull Request
- implement bootstrapper utility which embeds .msi installer
- refactor notification library
- CI


## PR Checklist
* [x] Applies to #4983


## Detailed Description of the Pull Request / Additional comments
Since we need to display toast notifications from the bootstrapper, and that requires it to have a shortcut, I've added dynamic icon creation functionality. However, Windows doesn't update its internal app list immediately, so we must sleep for 3s before showing the first notification. That will be fixed when we know how to register the app via registry (perhaps next Monday). Also, there's a known issue that we cannot determine if the bootstrapper was launched manually or during the update procedure, so for 0.20 the updating user will need to click through the msi dialogs. For later versions that's fixed by adding -silent flag during update.


## Validation Steps Performed
Preparation
- build my branch as v0.20 and publish it in my repo at
- [build 0.19.2 (1a2cd2a) with my repo test path as an update endpoint](https://github.com/yuyoyuppe/attachments/releases/download/v0.22/Yuyoyuppe-PowerToysSetup-0.19.2-x64.msi)

0.19.2 update via update now:
0.19.2 update via update at next launch:
0.19.2 elevated update via update now:
0.19.2 update via check for updates:
- works
- PT is updated

2) Launching the bootstrapper manually on a clean VM w/ no:
- dotnet is installed
- PT is installed
